libpromises/evalfunction.c

```diff
@@ -5119,69 +5119,85 @@ static FnCallResult FnCallFold(EvalContext *ctx,
 }
 
 /*********************************************************************/
-/* This function has been removed from the function list for now     */
-/*********************************************************************/
-#ifdef SUPPORT_FNCALL_DATATYPE
+
 static FnCallResult FnCallDatatype(EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
 {
-    const char* const varname = RlistScalarValue(finalargs);
+    assert(fp != NULL);
+    assert(fp->name != NULL);
+
+    if (finalargs == NULL)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Function %s requires variable identifier as first argument",
+            fp->name);
+        return FnFailure();
+    }
+    const char* const var_name = RlistScalarValue(finalargs);
 
-    VarRef* const ref = VarRefParse(varname);
+    VarRef* const var_ref = VarRefParse(var_name);
     DataType type;
-    const void *value = EvalContextVariableGet(ctx, ref, &type);
-    VarRefDestroy(ref);
+    const void *value = EvalContextVariableGet(ctx, var_ref, &type);
+    VarRefDestroy(var_ref);
 
-    Writer* const typestring = StringWriter();
+    /* detail argument defaults to false */
+    bool detail = false;
+    if (finalargs->next != NULL)
+    {
+        detail = BooleanFromString(RlistScalarValue(finalargs->next));
+    }
 
-    if (type == CF_DATA_TYPE_CONTAINER)
+    const char *const type_str = DataTypeToString(type);
+
+    if (!detail)
     {
+        return FnReturn(type_str);
+    }
 
-        const JsonElement* const jelement = value;
+    if (type == CF_DATA_TYPE_CONTAINER)
+    {
+        const char *subtype_str;
+        const JsonElement *const element = value;
 
-        if (JsonGetElementType(jelement) == JSON_ELEMENT_TYPE_CONTAINER)
+        if (JsonGetElementType(element) == JSON_ELEMENT_TYPE_CONTAINER)
```

Comment:
@larsewi Just use [`JsonGetType()`](https://github.com/cfengine/libntech/blob/8a22d70481deb4f93e41a1b58c416e0236b6a275/libutils/json.h#L184) instead, then you can make one big switch for [all the types](https://github.com/cfengine/libntech/blob/8a22d70481deb4f93e41a1b58c416e0236b6a275/libutils/json.h#L71-L80).